lib/testing/gpuSimulateCudaFRBBeamformer.cpp

```diff
@@ -0,0 +1,344 @@
+        S[i] = _dish_grid_size * _dishlayout[i * 2 + 0] + _dishlayout[i * 2 + 1];
+
+    while (!stop_thread) {
+        int4x2_t* voltage =
+            (int4x2_t*)wait_for_full_frame(voltage_buf, unique_name.c_str(), voltage_frame_id);
+        if (voltage == nullptr)
+            break;
+        float16_t* phase =
+            (float16_t*)wait_for_full_frame(phase_buf, unique_name.c_str(), phase_frame_id);
+        if (phase == nullptr)
+            break;
+        float16_t* output =
+            (float16_t*)wait_for_empty_frame(beamgrid_buf, unique_name.c_str(), beamgrid_frame_id);
+        if (output == nullptr)
+            break;
+
+        INFO("Simulating GPU processing for {:s}[{:d}] {:s}[{:d}] putting result in "
+             "{:s}[{:d}]",
+             voltage_buf->buffer_name, voltage_frame_id, phase_buf->buffer_name, phase_frame_id,
+             beamgrid_buf->buffer_name, beamgrid_frame_id);
+
+        std::string id_tag = std::to_string(voltage_frame_id);
+        if (metadata_is_onehot(voltage_buf, voltage_frame_id)) {
+            int frame_counter = get_onehot_frame_counter(voltage_buf, voltage_frame_id);
+            if (frame_counter < voltage_frame_id)
+                frame_counter = voltage_frame_id;
+            id_tag = std::to_string(frame_counter);
+        }
+
+        bool done = false;
+        DEBUG("Is voltage buffer one-hot? {:}  Phase? {:}",
+              metadata_is_onehot(voltage_buf, voltage_frame_id),
+              metadata_is_onehot(phase_buf, phase_frame_id));
+        if (metadata_is_onehot(voltage_buf, voltage_frame_id)) {
+            std::vector<int> inds = get_onehot_indices(voltage_buf, voltage_frame_id);
+            if (inds.size() == 0) {
+            } else if (inds.size() != 4) {
+                INFO("Expected 4 indices in voltage one-hot array, got {:d}", inds.size());
+            } else {
+                int t = inds[0];
+                int p = inds[1];
+                int f = inds[2];
+                int d = inds[3];
+                INFO("One-hot voltage buffer: time {:d} pol {:d}, freq {:d}, dish {:d}", t, p, f,
+                     d);
+#if KOTEKAN_FLOAT16
+                frb_simple_sub(S, phase, voltage, output, t, p, f, d);
+                done = true;
+#endif
+            }
+        }
+
+        if (!done && metadata_is_onehot(phase_buf, phase_frame_id)) {
+            std::vector<int> inds = get_onehot_indices(phase_buf, phase_frame_id);
+            if (inds.size() == 0) {
+            } else if (inds.size() != 5) {
+                INFO("Expected 5 indices in phase one-hot array, got {:d}", inds.size());
+            } else {
+                int f = inds[0];
+                int p = inds[1];
+                int b = inds[2];
+                int d = inds[3];
+                // real/imag = inds[4]
+                INFO("One-hot phase buffer: freq {:d} pol {:d}, beam {:d}, dish {:d}", f, p, b, d);
+                /////// DO WORK
+                // done = true;
```

Comment:
Removed this ... this was for doing a one-hot phase array, but for the FRB beamformer, that doesn't really work, because it has to compute the full "beamgrid" array which then gets interpolated to the actual beam locations.